scripts/build/types.js

```diff
@@ -6,8 +6,61 @@ const fs = require('fs');
 const path = require('path');
 const shell = require('shelljs');
 
+console.log(`Compiling definitions...`);
+
 shell.exec(`tsc -p tsconfig.declaration.json --outDir es/`);
 
+// replace block ts-ignore comments with line ones to support TS < 3.9
+shell.sed(
+  '-i',
+  /\*\* @ts-ignore/g,
+  '/ @ts-ignore',
+  path.join(__dirname, '../../es/**/*.d.ts')
+);
+
+// TypeScript uses `import()` types in very small files, but API extractor does not support that
+// We will hardcode the type definitions here, to avoid the build not passing.
+// "Error: The expression contains an import() type, which is not yet supported by API Extractor:"
+
+fs.writeFileSync(
+  path.join(
+    __dirname,
+    '../../es/connectors/infinite-hits/connectInfiniteHitsWithInsights.d.ts'
+  ),
+  `
+// edited by yarn build:types
+import { Connector } from '../../types';
+import {
+  InfiniteHitsRendererOptions,
+  InfiniteHitsConnectorParams,
+} from './connectInfiniteHits';
+
+declare const connectInfiniteHitsWithInsights: Connector<
+  InfiniteHitsRendererOptions,
+  InfiniteHitsConnectorParams
+>;
+export default connectInfiniteHitsWithInsights;
+`
+);
+
+fs.writeFileSync(
+  path.join(__dirname, '../../es/connectors/hits/connectHitsWithInsights.d.ts'),
+  `
+// edited by yarn build:types
+import { Connector } from '../../types';
+import {
+  HitsRendererOptions,
+  HitsConnectorParams,
+} from './connectHits';
+
+declare const connectHitsWithInsights: Connector<
+  HitsRendererOptions,
+  HitsConnectorParams
+>;
+export default connectHitsWithInsights;
+`
+);
+
```

Comment:
This seems very fragile to me since any change in the source file will not be reflected in the declaration file if we forget to update this code accordingly.

I think it would be better to change the source file to explicitly import the needed types, like they made here https://github.com/phryneas/react-async/commit/70393af94ddd030de56bfcadcd17f00e6762dfa9 (source: https://github.com/microsoft/rushstack/issues/1050#issuecomment-531656759)

We would get something like this for `connectInfiniteHitsWithInsights.ts`:

```ts
import { withInsights } from '../../lib/insights';
import connectInfiniteHits, {
  InfiniteHitsRendererOptions,
  InfiniteHitsConnectorParams,
} from './connectInfiniteHits';
import { Connector } from '../../types';

/**
 * Due to https://github.com/microsoft/web-build-tools/issues/1050, we need
 * Connector<...> imported in this file, even though it is only used implicitly.
 * This _uses_ Connector<...> so it is not accidentally removed by someone.
 */
declare type ImportWorkaround = Connector<
  InfiniteHitsRendererOptions,
  InfiniteHitsConnectorParams
>;

const connectInfiniteHitsWithInsights = withInsights(connectInfiniteHits);

export default connectInfiniteHitsWithInsights;
```

Reply:
That’s a good idea, I’ll try it